_posts/2023-04-21-tensorflow-mnist-classification.adoc

```diff
@@ -0,0 +1,405 @@
+If there are no new events within the specified time frame, the streaming finishes.
+
+In all cases obtained records are loaded into TensorFlow as the https://www.tensorflow.org/api_docs/python/tf/data/Dataset[Dataset].
+This is a convenient data structure provided by TensorFlow for building https://www.tensorflow.org/guide/data[data pipelines],  which may include further data transformations or preprocessing.
+
+This sounds great.
+However, the biggest caveat is how the records are represented within the Dataset.
+It seems that these Kafka loaders completely ignore the schema of the records provided by Kafka.
+Keys and values are always represented as the raw bytes.
+This won't be that bad, but even worse, it seems that somewhere on the way the data are converted into the strings (i.e. `toString()` method on the object is called).
+So if you pass e.g. raw image bytes via Kafka, using Kafka `BYTES_SCHEMA`, it would result into something like this:
+
+[source]
+----
+<tf.Tensor: shape=(64,), dtype=string, numpy=
+ array([b'[B@418b353d', b'[B@6aa28a4c', b'[B@b626485', b'[B@6d7491cd',
+        b'[B@13fa86c5', b'[B@7c3bc352', b'[B@64e5d61c', b'[B@2dd6d9b4',
+        b'[B@6addae65', b'[B@48ded13f', b'[B@2c1bb0e', b'[B@19c1d99b',
+        b'[B@1ee8f240', b'[B@20019f8b', b'[B@2f17494e', b'[B@380d4036',
+        b'[B@61aecf85', b'[B@4d7fe9fc', b'[B@58b79424', b'[B@ae963f4',
+        b'[B@1dac57cb', b'[B@2fae7d8b', b'[B@4b5ccaee', b'[B@aebf6b2',
+        b'[B@7506ea2b', b'[B@29989325', b'[B@43e2742', b'[B@51350f11',
+        b'[B@13a0f0ae', b'[B@7e4c4844', b'[B@b3d64f8', b'[B@7209bf09',
+        b'[B@66380466', b'[B@7aaa7e8d', b'[B@1ad0cf84', b'[B@259eca20',
+        b'[B@3a3f1c1', b'[B@36e4ff1f', b'[B@6578fc29', b'[B@79c924be',
+        b'[B@765b7f70', b'[B@67567aa3', b'[B@456d4bd4', b'[B@75317b13',
+        b'[B@58bc3a3a', b'[B@c6bc0ec', b'[B@2377095e', b'[B@5de017c0',
+        b'[B@64b48bac', b'[B@360a5b76', b'[B@2d2c9910', b'[B@70afd562',
+        b'[B@3006c930', b'[B@54b3e5ad', b'[B@1d1e0232', b'[B@1394d036',
+        b'[B@155dd43d', b'[B@5e88d5b6', b'[B@33ea53c7', b'[B@64a30ec',
+        b'[B@7dcdf024', b'[B@6570bf4e', b'[B@4e5bc4c', b'[B@537f216c'],
+       dtype=object)>,
+----
+
+Instead of getting a batch of raw image bytes which you can further transform in TensorFlow, you get only string representation of Java byte arrays, which is not very useful.
+
+To mitigate this issue, the most easy and straightforward approach seems to be to convert raw image bytes into numbers before sending them into Kafka.
+As TensorFlow provides methods for parsing CSV input, we can convert each image into one CSV line of numbers.
+TensorFlow mostly works with numbers, so we would have to convert our images into numbers anyway.
+The number which is on the image we can pass as a message key.
+Now, a single message transform supported by Debezium comes really handy.
+The transform can look like this:
```

Comment:
```suggestion
Since Tensorflow primarily works with numbers, we would be required to convert the images to numbers regardless.
We can pass the number on the image as a message key.
Now, a single message transform supported by Debezium comes in handy.
The transformation can look like this:
```